pkg/util/helper/mcs_test.go

```diff
@@ -0,0 +1,212 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	"github.com/karmada-io/karmada/pkg/util/gclient"
+)
+
+func TestCreateOrUpdateEndpointSlice(t *testing.T) {
+	type args struct {
+		client        client.Client
+		endpointSlice *discoveryv1.EndpointSlice
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *discoveryv1.EndpointSlice
+		wantErr bool
+	}{
+		{
+			name: "not exist and create",
+			args: args{
+				client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).Build(),
+				endpointSlice: &discoveryv1.EndpointSlice{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "eps", Namespace: "ns",
+						Labels: map[string]string{"foo": "foo1"},
+					},
+					AddressType: discoveryv1.AddressTypeIPv4,
+					Endpoints: []discoveryv1.Endpoint{{
+						Addresses: []string{"1.1.1.1"},
+					}},
+					Ports: []discoveryv1.EndpointPort{{
+						Port: pointer.Int32(80),
+					}},
+				},
+			},
+			want: &discoveryv1.EndpointSlice{
+				TypeMeta: metav1.TypeMeta{APIVersion: "discovery.k8s.io/v1", Kind: "EndpointSlice"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "eps", Namespace: "ns",
+					Labels: map[string]string{"foo": "foo1"},
+				},
+				AddressType: discoveryv1.AddressTypeIPv4,
+				Endpoints: []discoveryv1.Endpoint{{
+					Addresses:  []string{"1.1.1.1"},
+					Conditions: discoveryv1.EndpointConditions{},
+				}},
+				Ports: []discoveryv1.EndpointPort{{
+					Port: pointer.Int32(80),
+				}},
+			},
+			wantErr: false,
+		},
+		{
+			name: "exist and update",
+			args: args{
+				client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).WithObjects(
+					&discoveryv1.EndpointSlice{ObjectMeta: metav1.ObjectMeta{Name: "eps", Namespace: "ns"}}).Build(),
+				endpointSlice: &discoveryv1.EndpointSlice{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "eps", Namespace: "ns",
+						Labels: map[string]string{"foo": "foo1"},
+					},
+					AddressType: discoveryv1.AddressTypeIPv4,
+					Endpoints: []discoveryv1.Endpoint{{
+						Addresses: []string{"1.1.1.1"},
+					}},
+					Ports: []discoveryv1.EndpointPort{{
+						Port: pointer.Int32(80),
+					}},
+				},
+			},
+			want: &discoveryv1.EndpointSlice{
+				TypeMeta: metav1.TypeMeta{APIVersion: "discovery.k8s.io/v1", Kind: "EndpointSlice"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "eps", Namespace: "ns",
+					Labels: map[string]string{"foo": "foo1"},
+				},
+				AddressType: discoveryv1.AddressTypeIPv4,
+				Endpoints: []discoveryv1.Endpoint{{
+					Addresses:  []string{"1.1.1.1"},
+					Conditions: discoveryv1.EndpointConditions{},
+				}},
+				Ports: []discoveryv1.EndpointPort{{
+					Port: pointer.Int32(80),
+				}},
+			},
+			wantErr: false,
+		},
+		{
+			name: "exist and not update",
+			args: args{
+				client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).WithObjects(
```

Comment:
same with before, can we extract functions？

Reply:
Test objects have tiny differences.  Define it in the case, easy to read.